pkg/keycloak/keycloak.go

```diff
@@ -344,38 +354,60 @@ func (h *Handler) deleteKeycloak(kc *v1alpha1.Keycloak) error {
 }
 
 func (h *Handler) reconcileRealms(kc *v1alpha1.Keycloak, kcClient KeycloakInterface) error {
-	RealmPairsList := map[string]*KeycloakRealmPair{}
+	logrus.Debugf("reconcile realms, keycloak: %v", kc.Name)
+	realmPairsList := map[string]*KeycloakRealmPair{}
 
+	objRealms := kc.Spec.Realms
 	kcRealms, err := kcClient.ListRealms()
 	if err != nil {
 		return errors.Wrap(err, "error retrieving realms from keycloak")
 	}
 
-	for _, realm := range kc.Spec.Realms {
-		RealmPairsList[realm.Name] = &KeycloakRealmPair{
-			ObjRealm: &realm,
-			KcRealm:  kcRealms[realm.Name],
-		}
-		delete(kcRealms, realm.Name)
+	logrus.Debugf("kc realms: %#v, total: %v", kcRealms, len(kcRealms))
+	logrus.Debugf("obj realms: %#v, total: %v", objRealms, len(objRealms))
+
+	kcRealmMap := map[string]*v1alpha1.KeycloakRealm{}
+	for i := 0; i < len(kcRealms); i++ {
+		logrus.Debugf("kc realm %+v", kcRealms[i])
+		kcRealmMap[kcRealms[i].ID] = kcRealms[i]
 	}
 
-	for _, realm := range kcRealms {
-		RealmPairsList[realm.Name] = &KeycloakRealmPair{
-			KcRealm:  realm,
-			ObjRealm: nil,
+	for _, realm := range objRealms {
+		logrus.Debugf("obj realm %+v", realm)
+		realmPairsList[realm.ID] = &KeycloakRealmPair{
+			ObjRealm: &realm,
+			KcRealm:  kcRealmMap[realm.ID],
 		}
 	}
 
-	for name, realmPair := range RealmPairsList {
+	for name, realmPair := range realmPairsList {
 		err = h.reconcileRealm(realmPair.KcRealm, realmPair.ObjRealm, kcClient)
+		//This should try and reconcile all realms rather throwing an error on the first failure
 		if err != nil {
-			return errors.Wrap(err, "error reconciling realm "+name)
+			return errors.Wrapf(err, "error reconciling realm: %v", name)
 		}
 	}
 	return nil
```

Comment:
Took the decision here not to concern ourselves with realms that exist on Keycloak but not in the spec. I can imagine lots of scenarios where just deleting realms that we don't know about is goign to be a bad idea.

This actually makes "realmPairsList" defunct, and we are doing an extra iteration on the spec objects unnecessarily, but left it this way for now in case we change our minds about deleting realms (I don;t think we should :-)).